internal/servers/controller/handler.go

```diff
@@ -50,117 +52,88 @@ type HandlerProperties struct {
 	CancelCtx      context.Context
 }
 
-// Handler returns an http.Handler for the services. This can be used on
+// apiHandler returns an http.Handler for the services. This can be used on
 // its own to mount the Controller API within another web server.
-func (c *Controller) handler(props HandlerProperties) (http.Handler, error) {
-	// Create the muxer to handle the actual endpoints
+func (c *Controller) apiHandler(props HandlerProperties) (http.Handler, error) {
 	mux := http.NewServeMux()
 
-	h, err := handleGrpcGateway(c, props)
+	grpcGwMux := newGrpcGatewayMux()
+	err := registerGrpcGatewayEndpoints(props.CancelCtx, grpcGwMux, gatewayDialOptions(c.apiGrpcServerListener)...)
 	if err != nil {
 		return nil, err
 	}
-	mux.Handle("/v1/", h)
+	mux.Handle("/v1/", grpcGwMux)
 	mux.Handle("/", handleUi(c))
 
 	corsWrappedHandler := wrapHandlerWithCors(mux, props)
 	commonWrappedHandler := wrapHandlerWithCommonFuncs(corsWrappedHandler, c, props)
 	callbackInterceptingHandler := wrapHandlerWithCallbackInterceptor(commonWrappedHandler, c)
 	printablePathCheckHandler := cleanhttp.PrintablePathCheckHandler(callbackInterceptingHandler, nil)
 	eventsHandler, err := common.WrapWithEventsHandler(printablePathCheckHandler, c.conf.Eventer, c.kms, props.ListenerConfig)
-	if err != nil {
-		return nil, err
-	}
 
-	return eventsHandler, nil
+	return eventsHandler, err
 }
 
-func handleGrpcGateway(c *Controller, props HandlerProperties) (http.Handler, error) {
-	// Register*ServiceHandlerServer methods ignore the passed in ctx. Using it
-	// now however in case this changes in the future.
-	ctx := props.CancelCtx
-	currentServices := c.gatewayServer.GetServiceInfo()
-	dialOptions := gatewayDialOptions(c.gatewayListener)
+func (c *Controller) registerGrpcServices(ctx context.Context, s *grpc.Server) error {
+	// We have to check against the current services because the gRPC lib treats a duplicate
+	// register call as an error and os.Exits.
+	currentServices := s.GetServiceInfo()
 
 	if _, ok := currentServices[services.HostCatalogService_ServiceDesc.ServiceName]; !ok {
 		hcs, err := host_catalogs.NewService(c.StaticHostRepoFn, c.PluginHostRepoFn, c.HostPluginRepoFn, c.IamRepoFn)
 		if err != nil {
-			return nil, fmt.Errorf("failed to create host catalog handler service: %w", err)
-		}
-		services.RegisterHostCatalogServiceServer(c.gatewayServer, hcs)
-		if err := services.RegisterHostCatalogServiceHandlerFromEndpoint(ctx, c.gatewayMux, gatewayTarget, dialOptions); err != nil {
-			return nil, fmt.Errorf("failed to register host catalog service handler: %w", err)
```

Comment:
What is the rationale for moving these? I'm only wondering because it might make it easier to forget to both register the gRPC server and the gRPC-Gateway when adding new services, since they're in different functions now.

Reply:
Is there a reason we are starting one grpc-gateway object per api listener?